### ISSUE TITLE
IndexOutOfRangeException: When calling GetDescendant

### ISSUE THREAD

**Issue by ppasieka** [efcore/EFCore.SqlServer.HierarchyId#17]
I try to get a hierarchy id between two other ids. 

When I do:

```
	var hier = HierarchyId.Parse("/2/");

	var h1 = HierarchyId.Parse("/2/1/");
	var h2 = HierarchyId.Parse("/2/2/");

	var h1a = hier.GetDescendant(h1, h2); // -> /2/1.1/
	var h1aa = hier.GetDescendant(h1, h1a); // -> /2/1.0/    -------- this throws exception
	var h1aaa = hier.GetDescendant(h1aa, h1a); // -> /2/1.0.1/
```
Then I'm getting an exception.

I was pretty certain that it should work so I tried the same in MS SQL

```
DECLARE @h HIERARCHYID;
DECLARE @h1 HIERARCHYID;
DECLARE @h2 HIERARCHYID;
DECLARE @h1a HIERARCHYID;
DECLARE @h1aa HIERARCHYID;
DECLARE @h1aaa HIERARCHYID;

SET @h = '/2/';
SET @h1 = '/2/1/';
SET @h2 = '/2/2/';
SET @h1a = @h.GetDescendant(@h1, @h2);
SET @h1aa = @h.GetDescendant(@h1, @h1a);
SET @h1aaa = @h.GetDescendant(@h1aa, @h1a);

SELECT @h1aa.ToString()   --/2/1.0/
SELECT @h1aaa.ToString() --/2/1.0.1/

```
And it works there

**Comment by ppasieka** [comment]
I think I was too fast with creating this issue. It looks like problem is with the underlying implementation of the `SqlHierarchyId`

**Comment by bkarakaya01** [comment]
While I was trying to reproduce your issue, I've come up with another situation where:

```
var parent = HierarchyId.Parse("/1/");
var child1 = HierarchyId.Parse("/1/0.9/");
var child2 = HierarchyId.Parse("/1/1.1/");

var result = parent.GetDescendant(child1, child2);
```
Above, result will return **"0.1"**, I mean, if it's an expected result then there isn't any problem, but after reading the documents I was expecting **1.0**

**Comment by ppasieka** [comment]
Yes. I agree. There is a problem when calculating the hierarchyId.
The https://github.com/dotMorten/Microsoft.SqlServer.Types package used here is in v1.1 and it is a broken release. I tried to use the master branch from dotMorten repo and it fixes my exception, but it still calculates wrong ids in some cases. Luckily there is a PR which actually fixes all of the problems https://github.com/dotMorten/Microsoft.SqlServer.Types/pull/38 (I tried it and it seems to work :) )
